File: app/ui/src/main/java/com/fsck/k9/ui/helper/SizeFormatter.java
package com.fsck.k9.ui.helper;


import android.content.Context;

import com.fsck.k9.ui.R;

public class SizeFormatter {
    /*
     * Formats the given size as a String in bytes, kB, MB or GB with a single digit
     * of precision. Ex: 12,315,000 B = 12.3 MB
     */
    public static String formatSize(Context context, long size) {
        if (size > 1000000000) {
            return ((float)(size / 100000000) / 10) + context.getString(R.string.abbrev_gigabytes);
        }
        if (size > 1000000) {
            return ((float)(size / 100000) / 10) + context.getString(R.string.abbrev_megabytes);
        }
        if (size > 1000) {
            return ((float)(size / 100) / 10) + context.getString(R.string.abbrev_kilobytes);
        }
        return size + context.getString(R.string.abbrev_bytes);
    }

}

File: app/ui/src/main/java/com/fsck/k9/ui/helper/SizeFormatter.kt
package com.fsck.k9.ui.helper

import android.content.res.Resources
import com.fsck.k9.ui.R

class SizeFormatter(private val resources: Resources) {
    fun formatSize(size: Long): String {
        return when {
            size >= 999_950_000L -> {
                val number = (size / 1_000_000L).toFloat() / 1000f
                resources.getString(R.string.size_format_gigabytes, number)
            }
            size >= 999_950L -> {
                val number = (size / 1000L).toFloat() / 1000f
                resources.getString(R.string.size_format_megabytes, number)
            }
            size >= 1000L -> {
                val number = size.toFloat() / 1000f
                resources.getString(R.string.size_format_kilobytes, number)
            }
            else -> {
                resources.getString(R.string.size_format_bytes, size)
            }
        }
    }
}
